snapcraft/cli/lifecycle.py

```diff
@@ -22,23 +22,54 @@
 from . import echo
 from . import env
 from ._options import add_build_options, get_project
-from snapcraft.internal import deprecations, lifecycle, lxd, project_loader, steps
+from snapcraft.internal import (
+    build_providers,
+    deprecations,
+    lifecycle,
+    lxd,
+    project_loader,
+    steps,
+)
 from snapcraft.project.errors import YamlValidationError
 
 
-def _execute(step: steps.Step, parts, **kwargs):
-    project = get_project(**kwargs)
+# TODO: when snap is a real step we can simplify the arguments here.
+def _execute(step: steps.Step, parts, pack_project=False, **kwargs):
     build_environment = env.BuilderEnvironmentConfig()
+    project = get_project(is_managed_host=build_environment.is_managed_host, **kwargs)
 
-    if build_environment.is_host:
+    if project.info.base is not None and not (
+        build_environment.is_host or build_environment.is_managed_host
+    ):
+        build_provider_class = build_providers.get_provider_for(
+            build_environment.provider
+        )
+        echo.info("Launching a VM.")
+        with build_provider_class(project=project, echoer=echo) as instance:
+            instance.mount_project()
+            if pack_project:
+                # TODO add support for output
+                instance.pack_project(output=kwargs.get("output"))
+            else:
+                instance.execute_step(step)
+    elif build_environment.is_managed_host or build_environment.is_host:
```

Comment:
Yes, this is the traditional host build that we have today without bases. Or maybe clarify what your e.g. is for (i.e.; the if or the elif).

Reply:
It was for the elif. The elif can be hit both without a base, and WITH a base as long as is_host or is_managed_host. As long as both of those are okay, this is fine (although you just said this was the branch for building without bases, and it can be hit _with_ bases).

Reply:
let's cover the three scenarios
- if a base is set and this is not managed-host nor "nothing" (is_host) is set, then we setup a build provider, this build provider will rexecute snapcraft with managed-host set (the base is also set) running into the elif and correctly going through the lifecycle.
- if the base is not set then we do not setup a build environment and fall into the elif (is_host defaults to True).
- last, if is_lxd is set we go into that compatibility mode of the per project containers.